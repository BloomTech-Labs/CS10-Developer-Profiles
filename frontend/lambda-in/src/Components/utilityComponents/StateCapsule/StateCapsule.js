import React, { Component } from 'react';
import PropTypes from 'prop-types';

/**
 * Create a local state to manage an object's properties whilist is edited.
 *
 * @description
 */
class StateCapsule extends Component {
  constructor(props) {
    super(props);
    this.state = {
      ready: false,
    };
    this.removeItem = this.removeItem.bind(this);
    this.createItem = this.createItem.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleKeyPress = this.handleKeyPress.bind(this);
    // prettier-ignore
    this.mapAndResetKeysValues = object => Object.keys(object)
      .reduce((resetKeys, key) => {
        const valueType = Object.prototype.toString.call(object[key]);

        switch (valueType) {
          case '[object Array]':
            // eslint-disable-next-line no-param-reassign
            resetKeys[key] = [];
            // eslint-disable-next-line no-param-reassign
            resetKeys[`${key}_edit`] = '';
            break;
          default:
            // eslint-disable-next-line no-param-reassign
            resetKeys[key] = '';
        }

        return resetKeys;
      }, {});
  }

  componentDidMount() {
    this.resetState();
  }

  resetState() {
    const { schema, object } = this.props;
    const newState = {
      ...this.mapAndResetKeysValues(schema),
      ...object,
      ready: true,
    };

    // Set local state with object-properties and initialize its values.
    this.setState(newState);
  }

  /**
   * Sync local state with input field.
   */
  handleChange(event) {
    event.stopPropagation();

    const { id } = event.target;
    const details = id.split('-');
    const field = details[1];
    const { value } = event.target;

    this.setState({ [field]: value });
  }

  handleKeyPress(e) {
    e.stopPropagation();
    const { chips } = e.target.dataset;

    if (e.key === 'Enter' && chips) {
      e.preventDefault();

      const details = e.target.id.split('-');
      const { field } = e.target.dataset;
      const { value } = e.target;

      this.setState((prevState) => {
        const toUpdate = prevState[field];
        toUpdate.push(value);
        return { [field]: toUpdate, [details[1]]: '' };
      });
    }
  }

  createItem(field) {
    const newData = { ...this.state };
    delete newData.ready;

    const createEvent = new CustomEvent('onCreateItem', {
      bubbles: true,
      detail: {
        newData,
        field,
      },
    });
    return (e) => {
      e.stopPropagation();

      e.target.dispatchEvent(createEvent);
      this.resetState();
    };
  }

  removeItem(field, index) {
    const removeEvent = new CustomEvent('onDeleteItem', {
      bubbles: true,
      detail: { field, index },
    });

    return (e) => {
      e.stopPropagation();

      //
      e.target.dispatchEvent(removeEvent);

      const typeOfField = Object.prototype.toString.call(field);

      if (
        // prettier-ignore
        typeOfField === '[object String]'
        || typeOfField === '[object Number]'
        || typeOfField === '[object Boolean]'
      ) {
        this.setState((prevState) => {
          const toUpdate = prevState[field];
          toUpdate.splice(index, 1);
          return { [field]: toUpdate };
        });
      }
    };
  }

  render() {
    const { ready } = this.state;

    const stateCapsule = { ...this.state };
    delete stateCapsule.ready;

    const { children } = this.props;

    return ready ? (
      // eslint-disable-next-line jsx-a11y/no-static-element-interactions
      <div
        className="state-capsule"
        onChange={this.handleChange}
        onKeyPress={this.handleKeyPress}
      >
        {children({
          stateCapsule,
          removeItem: this.removeItem,
          createItem: this.createItem,
        })}
      </div>
    ) : null;
  }
}

StateCapsule.propTypes = {
  schema: PropTypes.shape({}).isRequired,
  object: PropTypes.shape({}).isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
    PropTypes.func,
  ]).isRequired,
};

export default StateCapsule;
