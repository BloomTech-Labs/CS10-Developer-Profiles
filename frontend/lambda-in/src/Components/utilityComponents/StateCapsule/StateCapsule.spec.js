import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import StateCapsule from './StateCapsule';

configure({ adapter: new Adapter() });

/**
 * @author jesuarva
 */
describe('StateCapsule component', () => {
  const schema = {
    forRemoveButton: {
      summary: 'Summary',
      topSkills: ['Top Skills'],
    },
    forCreateButton: {
      title: 'Title',
      description: 'Description',
      img: 'Image',
      link: 'Link',
      repo: 'Repository',
      // tech: ['Stack'], // TODO
    },
  };

  const data = {
    topSkills: ['JS', 'React', 'TDD', 'Full stack developer'],
    summary: 'Some summary here',
  };

  const stateCapsules = {
    basic: (
      <StateCapsule schema={schema.forRemoveButton} object={data}>
        {() => {}}
      </StateCapsule>
    ),
    withChildrenRemoveButton: (
      <StateCapsule schema={schema.forRemoveButton} object={data}>
        {({ stateCapsule, removeItem }) => (
          <div>
            <input id="edit-summary" type="text" value={stateCapsule.summary} />
            <input
              id="edit-topSkills_edit"
              type="text"
              value={stateCapsule.summary}
            />
            <ul>
              {stateCapsule.topSkills.map((skill, index) => (
                <div key={`${Math.random()}-${Date.now()}`}>
                  <li>
                    {skill}
                    <button
                      className="button-remove"
                      type="button"
                      onClick={removeItem('topSkills', index)}
                    >
                      X
                    </button>
                  </li>
                </div>
              ))}
            </ul>
          </div>
        )}
      </StateCapsule>
    ),
    withChildrenCreateButton: (
      <StateCapsule schema={schema.forCreateButton} object={{}}>
        {({ stateCapsule, createItem }) => (
          <div>
            <button
              className="button-create"
              type="button"
              onClick={createItem('projects')}
            >
              Create
            </button>
            <ul>
              {// prettier-ignore
              Object.keys(stateCapsule).map(field => (
                <div key={`${Math.random()}-${Date.now()}`}>
                  <input
                    id={`new-${field}`}
                    type="text"
                    value={stateCapsule[field]}
                  />
                </div>
              ))}
            </ul>
          </div>
        )}
      </StateCapsule>
    ),
  };

  describe('Renders correctly', () => {
    it('should renders basic component', () => {
      const tree = renderer.create(stateCapsules.basic).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should renders children with remove functionality.', () => {
      const tree = renderer
        .create(stateCapsules.withChildrenRemoveButton)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should renders children with create functionality.', () => {
      const tree = renderer
        .create(stateCapsules.withChildrenCreateButton)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  it('should initialize state with schema`s properties', () => {
    const shallowSC = shallow(stateCapsules.basic);

    const scState = shallowSC.state();

    expect(scState).toHaveProperty('summary', 'Some summary here');
    expect(scState).toHaveProperty('topSkills_edit');
    expect(scState).toHaveProperty('topSkills', [
      'JS',
      'React',
      'TDD',
      'Full stack developer',
    ]);
  });

  it('should update state onChange', () => {
    const shallowSC = shallow(stateCapsules.withChildrenRemoveButton);

    // Before any 'onChange' event
    expect(shallowSC.state().topSkills_edit).toMatch('');

    shallowSC.find('.state-capsule').simulate('change', {
      target: {
        id: 'edit-topSkills_edit',
        value: 'JavaScrip',
      },
      stopPropagation: () => {},
    });

    // After an 'onChange' event
    expect(shallowSC.state().topSkills_edit).toMatch('JavaScrip');
  });

  describe('Remove items from an array', () => {
    let shallowSC;

    // Initialized Array = ['JS', 'React', 'TDD', 'Full stack developer'];
    const modifiedArray = [];

    const mockDispatchEvent = jest.fn(() => {});
    const mockStopPropagation = jest.fn(() => {});
    // prettier-ignore
    const mockEvent = index => Object.assign({
      target: {
        field: 'topSkills',
        index,
        dispatchEvent: mockDispatchEvent,
      },
      stopPropagation: mockStopPropagation,
    });

    beforeAll(() => {
      shallowSC = shallow(stateCapsules.withChildrenRemoveButton);

      const buttons = shallowSC.find('.button-remove');

      /**
       * Simulate 'click' on each rendered button, thus all items are removed.
       */
      let aux = 0; // to mimic an index
      buttons.forEach((button) => {
        button.simulate('click', mockEvent(aux));
        aux += 1;
      });
    });

    it('should update state`s array-like property onClick', () => {
      // After on click event => array must be empty
      expect(shallowSC.state().topSkills).toEqual(
        expect.arrayContaining(modifiedArray),
      );
    });

    it('should call dispatchEvent', () => {
      expect(mockDispatchEvent).toHaveBeenCalledTimes(4);
    });

    it('should call stopPropagation', () => {
      expect(mockStopPropagation).toHaveBeenCalledTimes(4);
    });
  });

  describe('Create new array-item and push it to its array', () => {
    let shallowSC;
    let button;
    const mockDispatchEvent = jest.fn(() => {});
    const mockStopPropagation = jest.fn(() => {});
    const mockEvent = {
      target: {
        field: 'projects',
        dispatchEvent: mockDispatchEvent,
      },
      stopPropagation: mockStopPropagation,
    };

    beforeAll(() => {
      shallowSC = shallow(stateCapsules.withChildrenCreateButton);

      button = shallowSC.find('button.button-create');

      /**
       * Simulate user interaction typing some text in input fields.
       */
      shallowSC.setState({
        title: 'A happy user filling title field',
        description: 'A happy user filling description field',
        img: 'A happy user filling img field',
        link: 'A happy user filling link field',
        repo: 'A happy user filling repo field',
      });

      /**
       * Submit the form and create a new entry in the corresponding array.
       *
       * This user interaction shall also reset the state.
       */
      button.simulate('click', mockEvent);
    });

    it('should have only one create-button', () => {
      expect(button).toHaveLength(1);
    });

    it('should call dispatchEvent', () => {
      expect(mockDispatchEvent).toHaveBeenCalledTimes(1);
    });

    it('should call stopPropagation', () => {
      expect(mockStopPropagation).toHaveBeenCalledTimes(1);
    });

    it('should reset state', () => {
      expect(shallowSC.state()).toMatchObject({
        title: '',
        description: '',
        img: '',
        link: '',
        repo: '',
      });
    });
  });
});
