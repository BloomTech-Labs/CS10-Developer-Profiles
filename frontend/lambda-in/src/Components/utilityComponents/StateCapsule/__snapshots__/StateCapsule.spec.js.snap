// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StateCapsule component Renders correctly should renders basic component 1`] = `
<div
  className="state-capsule"
  onChange={[Function]}
  onKeyPress={[Function]}
/>
`;

exports[`StateCapsule component Renders correctly should renders children with create functionality. 1`] = `
<div
  className="state-capsule"
  onChange={[Function]}
  onKeyPress={[Function]}
>
  <div>
    <button
      className="button-create"
      onClick={[Function]}
      type="button"
    >
      Create
    </button>
    <ul>
      <div>
        <input
          id="new-title"
          type="text"
          value=""
        />
      </div>
      <div>
        <input
          id="new-description"
          type="text"
          value=""
        />
      </div>
      <div>
        <input
          id="new-img"
          type="text"
          value=""
        />
      </div>
      <div>
        <input
          id="new-link"
          type="text"
          value=""
        />
      </div>
      <div>
        <input
          id="new-repo"
          type="text"
          value=""
        />
      </div>
    </ul>
  </div>
</div>
`;

exports[`StateCapsule component Renders correctly should renders children with remove functionality. 1`] = `
<div
  className="state-capsule"
  onChange={[Function]}
  onKeyPress={[Function]}
>
  <div>
    <input
      id="edit-summary"
      type="text"
      value="Some summary here"
    />
    <input
      id="edit-topSkills_edit"
      type="text"
      value="Some summary here"
    />
    <ul>
      <div>
        <li>
          JS
          <button
            className="button-remove"
            onClick={[Function]}
            type="button"
          >
            X
          </button>
        </li>
      </div>
      <div>
        <li>
          React
          <button
            className="button-remove"
            onClick={[Function]}
            type="button"
          >
            X
          </button>
        </li>
      </div>
      <div>
        <li>
          TDD
          <button
            className="button-remove"
            onClick={[Function]}
            type="button"
          >
            X
          </button>
        </li>
      </div>
      <div>
        <li>
          Full stack developer
          <button
            className="button-remove"
            onClick={[Function]}
            type="button"
          >
            X
          </button>
        </li>
      </div>
    </ul>
  </div>
</div>
`;
